sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala

```diff
@@ -2457,164 +2450,127 @@ class Analyzer(override val catalogManager: CatalogManager)
+          // The given expression is valid and doesn't need extra resolution.
+          maybeResolved
+        } else if (containsUnresolvedFunc(maybeResolved)) {
+          // The given expression has unresolved functions which may be aggregate functions and we
+          // need to wait for other rules to resolve the functions first.
+          maybeResolved
+        } else {
+          // Avoid adding an extra aggregate expression if it's already present in
+          // `agg.aggregateExpressions`.
+          val index = if (maybeResolved.resolved) {
+            agg.aggregateExpressions.indexWhere {
+              case Alias(child, _) => child semanticEquals maybeResolved
+              case other => other semanticEquals maybeResolved
+            }
           } else {
-            Project(aggregate.output,
-              Sort(finalSortOrders, global,
-                aggregate.copy(aggregateExpressions = originalAggExprs ++ needsPushDown)))
+            -1
+          }
+          if (index >= 0) {
+            agg.aggregateExpressions(index).toAttribute
+          } else {
+            buildAggExprList(maybeResolved, agg, aggregateExpressions)
           }
-        } catch {
-          // Attempting to resolve in the aggregate can result in ambiguity.  When this happens,
-          // just return the original plan.
-          case ae: AnalysisException => sort
         }
+      }
+      (aggregateExpressions.toSeq, transformed)
     }
 
-    def hasCharVarchar(expr: Alias): Boolean = {
-      expr.find {
-        case ne: NamedExpression => CharVarcharUtils.getRawType(ne.metadata).nonEmpty
-        case _ => false
-      }.nonEmpty
+    private def buildAggExprList(
+        expr: Expression,
+        agg: Aggregate,
+        aggExprList: ArrayBuffer[NamedExpression]): Expression = expr match {
+      case ae: AggregateExpression if ae.resolved =>
+        val alias = Alias(ae, ae.toString)()
+        aggExprList += alias
+        alias.toAttribute
+      // Grouping functions are handled in the rule [[ResolveGroupingAnalytics]].
+      case grouping: Expression if grouping.resolved &&
+          agg.groupingExpressions.exists(_.semanticEquals(grouping)) &&
+          !ResolveGroupingAnalytics.hasGroupingFunction(grouping) &&
+          !agg.output.exists(_.semanticEquals(grouping)) =>
+        grouping match {
+          case ne: NamedExpression =>
+            aggExprList += ne
+            ne.toAttribute
+          case _ =>
+            val alias = Alias(grouping, grouping.toString)()
+            aggExprList += alias
+            alias.toAttribute
+        }
+      case a: Attribute if agg.child.outputSet.contains(a) && !agg.outputSet.contains(a) =>
+        // Undo the resolution. This attribute is neither inside aggregate functions nor a
+        // grouping column. It shouldn't be resolved with `agg.child.output`.
+        CurrentOrigin.withOrigin(a.origin)(UnresolvedAttribute(Seq(a.name)))
```

Comment:
Just out of curiosity; what's this handling for?

Reply:
Some tests failed because the error message is changed. This is to restore the previous error message, which is more pretty.